data race in scan UT

16:55:43 === RUN TestCheckerTestSuite/TestIsScannable
16:55:43 ==================
16:55:43 WARNING: DATA RACE
16:55:43 Read at 0x0000019624c0 by goroutine 52:
16:55:43 github.com/goharbor/harbor/src/controller/artifact.Iterator.func1()
16:55:43 /home/runner/work/harbor/harbor/src/github.com/goharbor/harbor/src/controller/artifact/helper.go:36 +0x239
16:55:43
16:55:43 Previous write at 0x0000019624c0 by goroutine 49:
16:55:43 github.com/goharbor/harbor/src/controller/scan.(*CallbackTestSuite).TearDownSuite()
16:55:43 /home/runner/work/harbor/harbor/src/github.com/goharbor/harbor/src/controller/scan/callback_test.go:92 +0xcb
16:55:43 github.com/goharbor/harbor/src/vendor/github.com/stretchr/testify/suite.Run.func2()
16:55:43 /home/runner/work/harbor/harbor/src/github.com/goharbor/harbor/src/vendor/github.com/stretchr/testify/suite/suite.go:166 +0x1be
16:55:43 github.com/goharbor/harbor/src/vendor/github.com/stretchr/testify/suite.Run()
16:55:43 /home/runner/work/harbor/harbor/src/github.com/goharbor/harbor/src/vendor/github.com/stretchr/testify/suite/suite.go:177 +0x925
16:55:43 github.com/goharbor/harbor/src/controller/scan.TestCallbackTestSuite()
16:55:43 /home/runner/work/harbor/harbor/src/github.com/goharbor/harbor/src/controller/scan/callback_test.go:221 +0xb1
16:55:43 testing.tRunner()
16:55:43 /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1123 +0x202
16:55:43
16:55:43 Goroutine 52 (running) created at:
16:55:43 github.com/goharbor/harbor/src/controller/artifact.Iterator()
16:55:43 /home/runner/work/harbor/harbor/src/github.com/goharbor/harbor/src/controller/artifact/helper.go:28 +0xb8
16:55:43 github.com/goharbor/harbor/src/controller/scan.(*basicController).startScanAll()
16:55:43 /home/runner/work/harbor/harbor/src/github.com/goharbor/harbor/src/controller/scan/base_controller.go:309 +0xa8
16:55:43 github.com/goharbor/harbor/src/controller/scan.(*basicController).ScanAll.func1()
16:55:43 /home/runner/work/harbor/harbor/src/github.com/goharbor/harbor/src/controller/scan/base_controller.go:293 +0x1e8
16:55:43
16:55:43 Goroutine 49 (finished) created at:
16:55:43 testing.(*T).Run()
16:55:43 /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1168 +0x5bb
16:55:43 testing.runTests.func1()
16:55:43 /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1439 +0xa6
16:55:43 testing.tRunner()
16:55:43 /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1123 +0x202
16:55:43 testing.runTests()
16:55:43 /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1437 +0x612
16:55:43 testing.(*M).Run()
16:55:43 /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1345 +0x3b3
16:55:43 main.main()
16:55:43 _testmain.go:411 +0x356
16:55:43 ==================
Error: 3 2021-01-03T16:55:43Z [ERROR] [/github.com/goharbor/harbor/src/controller/artifact/helper.go:38]: list artifacts failed, error: cannot get the ORM from context
16:55:43 === RUN TestCheckerTestSuite/TestScannerNotFound
16:55:43 === CONT TestCheckerTestSuite
16:55:43 testing.go:1038: race detected during execution of test
16:55:43 --- FAIL: TestCheckerTestSuite (0.01s)
16:55:43 --- PASS: TestCheckerTestSuite/TestIsScannable (0.00s)
16:55:43 --- PASS: TestCheckerTestSuite/TestScannerNotFound (0.00s)
16:55:43 === CONT
16:55:43 testing.go:1038: race detected during execution of test
16:55:43 FAIL

--- a/src/controller/scan/callback.go
+++ b/src/controller/scan/callback.go
@@ -34,6 +34,13 @@ const (
 	ScanAllCallback = "scanAll"
 )
 
+var (
+	artifactCtl = artifact.Ctl
+	robotCtl    = robot.Ctl
+	scanCtl     = DefaultController
+	taskMgr     = task.Mgr
+)
+
 func init() {
 	if err := scheduler.RegisterCallbackFunc(ScanAllCallback, scanAllCallback); err != nil {
 		log.Fatalf("failed to register the callback for the scan all schedule, error %v", err)
@@ -54,7 +61,7 @@ func init() {
 }
 
 func scanAllCallback(ctx context.Context, param string) error {
-	_, err := DefaultController.ScanAll(ctx, task.ExecutionTriggerSchedule, true)
+	_, err := scanCtl.ScanAll(ctx, task.ExecutionTriggerSchedule, true)
 	return err
 }
 
@@ -64,7 +71,7 @@ func scanTaskStatusChange(ctx context.Context, taskID int64, status string) (err
 	js := job.Status(status)
 
 	if js.Final() {
-		t, err := task.Mgr.Get(ctx, taskID)
+		t, err := taskMgr.Get(ctx, taskID)
 		if err != nil {
 			return err
 		}
@@ -72,7 +79,7 @@ func scanTaskStatusChange(ctx context.Context, taskID int64, status string) (err
 		if js == job.SuccessStatus {
 			robotID := getRobotID(t.ExtraAttrs)
 			if robotID > 0 {
-				if err := robot.Ctl.Delete(ctx, robotID); err != nil {
+				if err := robotCtl.Delete(ctx, robotID); err != nil {
 					// Should not block the main flow, just logged
 					logger.WithFields(log.Fields{"robot_id": robotID, "error": err}).Error("delete robot account failed")
 				} else {
@@ -83,7 +90,7 @@ func scanTaskStatusChange(ctx context.Context, taskID int64, status string) (err
 
 		artifactID := getArtifactID(t.ExtraAttrs)
 		if artifactID > 0 {
-			art, err := artifact.Ctl.Get(ctx, artifactID, nil)
+			art, err := artifactCtl.Get(ctx, artifactID, nil)
 			if err != nil {
 				logger.WithFields(log.Fields{"artifact_id": artifactID, "error": err}).Errorf("failed to get artifact")
 			} else {
@@ -114,5 +121,5 @@ func scanTaskCheckInProcessor(ctx context.Context, t *task.Task, data string) (e
 		return err
 	}
 
-	return DefaultController.UpdateReport(ctx, checkInReport)
+	return scanCtl.UpdateReport(ctx, checkInReport)
 }

--- a/src/controller/scan/callback_test.go
+++ b/src/controller/scan/callback_test.go
@@ -18,8 +18,9 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"testing"
+
 	"github.com/goharbor/harbor/src/controller/artifact"
-	"github.com/goharbor/harbor/src/controller/robot"
 	"github.com/goharbor/harbor/src/jobservice/job"
 	"github.com/goharbor/harbor/src/lib/q"
 	"github.com/goharbor/harbor/src/pkg/scan"
@@ -32,48 +33,39 @@ import (
 	reporttesting "github.com/goharbor/harbor/src/testing/pkg/scan/report"
 	tasktesting "github.com/goharbor/harbor/src/testing/pkg/task"
 	"github.com/stretchr/testify/suite"
-	"testing"
 )
 
 type CallbackTestSuite struct {
 	suite.Suite
 
-	artifactCtl         *artifacttesting.Controller
-	originalArtifactCtl artifact.Controller
+	artifactCtl *artifacttesting.Controller
 
 	execMgr *tasktesting.ExecutionManager
 
-	robotCtl         *robottesting.Controller
-	originalRobotCtl robot.Controller
+	robotCtl *robottesting.Controller
 
 	reportMgr *reporttesting.Manager
 
-	scanCtl         Controller
-	originalScanCtl Controller
+	scanCtl Controller
 
 	taskMgr         *tasktesting.Manager
-	originalTaskMgr task.Manager
 	reportConverter *postprocessorstesting.ScanReportV1ToV2Converter
 }
 
 func (suite *CallbackTestSuite) SetupSuite() {
-	suite.originalArtifactCtl = artifact.Ctl
 	suite.artifactCtl = &artifacttesting.Controller{}
-	artifact.Ctl = suite.artifactCtl
+	artifactCtl = suite.artifactCtl
 
 	suite.execMgr = &tasktesting.ExecutionManager{}
 
-	suite.originalRobotCtl = robot.Ctl
 	suite.robotCtl = &robottesting.Controller{}
-	robot.Ctl = suite.robotCtl
+	robotCtl = suite.robotCtl
 
 	suite.reportMgr = &reporttesting.Manager{}
 
-	suite.originalTaskMgr = task.Mgr
 	suite.taskMgr = &tasktesting.Manager{}
-	task.Mgr = suite.taskMgr
+	taskMgr = suite.taskMgr
 
-	suite.originalScanCtl = DefaultController
 	suite.reportConverter = &postprocessorstesting.ScanReportV1ToV2Converter{}
 
 	suite.scanCtl = &basicController{
@@ -83,15 +75,7 @@ func (suite *CallbackTestSuite) SetupSuite() {
 		taskMgr:         suite.taskMgr,
 		reportConverter: suite.reportConverter,
 	}
-	DefaultController = suite.scanCtl
-}
-
-func (suite *CallbackTestSuite) TearDownSuite() {
-	DefaultController = suite.originalScanCtl
-
-	artifact.Ctl = suite.originalArtifactCtl
-	robot.Ctl = suite.originalRobotCtl
-	task.Mgr = suite.originalTaskMgr
+	scanCtl = suite.scanCtl
 }
 
 func (suite *CallbackTestSuite) TestScanTaskStatusChange() {

